File: frontend/src/tools/bank_journal/api/customFields.ts
import { apiClient } from '../../../api/client';

/** 标准字段定义（内置 + 公司扩展合并视图）。 */
export interface StandardFieldDef {
  key: string;
  label: string;
  type: 'text' | 'amount' | 'date' | 'enum';
  /** True=内置字段，False=公司扩展字段 */
  builtin: boolean;
  /** 当前生效识别关键词（内置默认 ∪ 公司覆盖） */
  keywords: string[];
  /** 内置字段是否有公司覆盖（前端用于显示「恢复默认」） */
  overridden: boolean;
}

export interface SlotQuota {
  text: { used: number; total: number };
  amount: { used: number; total: number };
  date: { used: number; total: number };
}

export interface StandardSchema {
  fields: StandardFieldDef[];
  slot_quota: SlotQuota;
}

export interface CustomField {
  id: string;
  company_id: string;
  field_key: string;
  name: string;
  slot_key: string;
  data_type: 'text' | 'amount' | 'date';
  header_keywords: string[];
  status: string;
}

export interface CustomFieldCreate {
  company_id: string;
  field_key: string;
  name: string;
  data_type: 'text' | 'amount' | 'date';
  header_keywords: string[];
  created_by?: string;
}

export interface CustomFieldUpdate {
  name?: string;
  header_keywords?: string[];
  status?: 'active' | 'inactive';
}

const BASE = '/api/tools/bank-journal/custom-fields';

/** 合并后的标准字段全集（内置 + 该公司扩展），供所有字段下拉运行时拉取。 */
export async function getStandardSchema(companyId: string): Promise<StandardSchema> {
  const response = await apiClient.get<StandardSchema>(`${BASE}/standard-schema`, {
    params: { company_id: companyId }
  });
  return response.data;
}

export async function listCustomFields(companyId?: string): Promise<CustomField[]> {
  const params = companyId ? { company_id: companyId } : undefined;
  const response = await apiClient.get<CustomField[]>(BASE, { params });
  return response.data;
}

export async function createCustomField(payload: CustomFieldCreate): Promise<CustomField> {
  const response = await apiClient.post<CustomField>(BASE, payload);
  return response.data;
}

export async function updateCustomField(
  id: string,
  payload: CustomFieldUpdate
): Promise<CustomField> {
  const response = await apiClient.patch<CustomField>(`${BASE}/${id}`, payload);
  return response.data;
}

export async function deleteCustomField(id: string): Promise<void> {
  await apiClient.delete(`${BASE}/${id}`);
}

// ---------------------------------------------------------------------------
// 内置字段覆盖（公司级：label / 识别关键词 / 规则类型）
// ---------------------------------------------------------------------------

export interface BuiltinFieldOverride {
  id: string;
  company_id: string;
  field_key: string;
  label_override: string | null;
  header_keywords_override: string[] | null;
  type_override: 'text' | 'amount' | 'date' | 'enum' | null;
}

export interface BuiltinOverrideUpsert {
  company_id: string;
  field_key: string;
  label_override?: string | null;
  header_keywords_override?: string[] | null;
  type_override?: 'text' | 'amount' | 'date' | 'enum' | null;
}

/** upsert 内置字段覆盖（field_key 必须是内置字段）。返回公司级覆盖列表。 */
export async function listBuiltinOverrides(companyId: string): Promise<BuiltinFieldOverride[]> {
  const response = await apiClient.get<BuiltinFieldOverride[]>(`${BASE}/builtin-overrides`, {
    params: { company_id: companyId }
  });
  return response.data;
}

export async function upsertBuiltinOverride(
  payload: BuiltinOverrideUpsert
): Promise<BuiltinFieldOverride> {
  const response = await apiClient.put<BuiltinFieldOverride>(
    `${BASE}/builtin-overrides/${payload.field_key}`,
    payload
  );
  return response.data;
}

/** 删除覆盖即恢复内置默认。 */
export async function deleteBuiltinOverride(
  fieldKey: string,
  companyId: string
): Promise<void> {
  await apiClient.delete(`${BASE}/builtin-overrides/${fieldKey}`, {
    params: { company_id: companyId }
  });
}

/**
 * 供编辑器下拉使用的字段选项 + 类型映射，运行时从后端拉取（含公司扩展字段）。
 *
 * 返回 {
 *   options:  [{value,label}]  —— 注入 MappingEditor/RuleEditor/DetectResultView 的 standardFieldOptions
 *   typeMap:  {key→type}       —— 注入 RuleEditor 的 fieldTypeMap（条件操作符智能过滤）
 *   loading
 * }
 * 失败时回退到内置字段（由组件自身默认值兜底），不阻断页面。
 */
export interface StandardFieldOptionsState {
  options: { value: string; label: string }[];
  typeMap: Record<string, string>;
  loading: boolean;
}

export function standardFieldOptionsFromSchema(
  schema: StandardSchema | null
): { options: { value: string; label: string }[]; typeMap: Record<string, string> } {
  if (!schema) return { options: [], typeMap: {} };
  const options = schema.fields.map((f) => ({ value: f.key, label: f.label }));
  const typeMap: Record<string, string> = {};
  for (const f of schema.fields) typeMap[f.key] = f.type;
  return { options, typeMap };
}

File: backend/app/tools/bank_journal/routes/custom_fields.py
"""公司级自定义扩展字段的 HTTP 路由。"""

from fastapi import APIRouter, Depends

from app.api.deps import (
    CurrentUserDep,
    DbSession,
    accessible_company_filter,
    require,
    require_company_access,
)
from app.core.permissions import Permission
from app.tools.bank_journal.models.custom_field import CustomField
from app.tools.bank_journal.schemas.custom_field import (
    BuiltinFieldOverrideResponse,
    BuiltinFieldOverrideUpsert,
    CustomFieldCreate,
    CustomFieldResponse,
    CustomFieldUpdate,
    StandardSchemaResponse,
)
from app.tools.bank_journal.services import custom_field_service

router = APIRouter(prefix="/api/tools/bank-journal/custom-fields", tags=["custom-fields"])


@router.get(
    "/standard-schema",
    response_model=StandardSchemaResponse,
    dependencies=[Depends(require(Permission.READ))],
)
def get_standard_schema(
    db: DbSession, user: CurrentUserDep, company_id: str
) -> StandardSchemaResponse:
    """返回内置标准字段（含公司覆盖）+ 公司扩展字段的合并视图，供前端字段下拉运行时拉取。"""
    require_company_access(user, company_id)  # 跨公司读取拦截
    return custom_field_service.get_standard_schema(db, company_id)


@router.get(
    "",
    response_model=list[CustomFieldResponse],
    dependencies=[Depends(require(Permission.READ))],
)
def list_custom_fields(
    db: DbSession, user: CurrentUserDep, company_id: str | None = None
) -> list[CustomFieldResponse]:
    return custom_field_service.list_custom_fields(
        db, company_id, accessible=accessible_company_filter(user)
    )


@router.post(
    "",
    response_model=CustomFieldResponse,
    status_code=201,
    dependencies=[Depends(require(Permission.TEMPLATE_MANAGE))],
)
def create_custom_field(
    db: DbSession, user: CurrentUserDep, payload: CustomFieldCreate
) -> CustomFieldResponse:
    require_company_access(user, payload.company_id)
    payload.created_by = user.id
    return custom_field_service.create_custom_field(db, payload)


@router.patch(
    "/{field_id}",
    response_model=CustomFieldResponse,
    dependencies=[Depends(require(Permission.TEMPLATE_MANAGE))],
)
def update_custom_field(
    db: DbSession, user: CurrentUserDep, field_id: str, payload: CustomFieldUpdate
) -> CustomFieldResponse:
    _require_field_company_access(db, user, field_id)
    return custom_field_service.update_custom_field(db, field_id, payload)


@router.delete(
    "/{field_id}",
    status_code=204,
    dependencies=[Depends(require(Permission.TEMPLATE_MANAGE))],
)
def delete_custom_field(db: DbSession, user: CurrentUserDep, field_id: str) -> None:
    _require_field_company_access(db, user, field_id)
    custom_field_service.delete_custom_field(db, field_id)


def _require_field_company_access(db: DbSession, user: CurrentUserDep, field_id: str) -> None:
    """加载扩展字段并按其公司做写访问校验（不存在则保持 404 由 service 抛出）。"""
    cf = db.query(CustomField).filter(CustomField.id == field_id).first()
    if cf is None:
        return  # 交由 service 层 _get_or_404 抛 404，避免把 404 变 403
    require_company_access(user, cf.company_id)


# ---------------------------------------------------------------------------
# 内置字段覆盖（公司级：label / 识别关键词 / 规则类型）
# ---------------------------------------------------------------------------


@router.get(
    "/builtin-overrides",
    response_model=list[BuiltinFieldOverrideResponse],
    dependencies=[Depends(require(Permission.READ))],
)
def list_builtin_overrides(
    db: DbSession, user: CurrentUserDep, company_id: str
) -> list[BuiltinFieldOverrideResponse]:
    require_company_access(user, company_id)  # 跨公司读取拦截
    return custom_field_service.list_builtin_overrides(db, company_id)


@router.put(
    "/builtin-overrides/{field_key}",
    response_model=BuiltinFieldOverrideResponse,
    dependencies=[Depends(require(Permission.TEMPLATE_MANAGE))],
)
def upsert_builtin_override(
    db: DbSession,
    user: CurrentUserDep,
    field_key: str,
    payload: BuiltinFieldOverrideUpsert,
) -> BuiltinFieldOverrideResponse:
    require_company_access(user, payload.company_id)
    # path 参数与 body 的 field_key 必须一致
    payload.field_key = field_key
    return custom_field_service.upsert_builtin_override(db, payload)


@router.delete(
    "/builtin-overrides/{field_key}",
    status_code=204,
    dependencies=[Depends(require(Permission.TEMPLATE_MANAGE))],
)
def delete_builtin_override(
    db: DbSession, user: CurrentUserDep, field_key: str, company_id: str
) -> None:
    require_company_access(user, company_id)
    custom_field_service.delete_builtin_override(db, company_id, field_key)
